keps/sig-apps/19-Graduate-CronJob-to-Stable/README.md

```diff
@@ -0,0 +1,503 @@
+// It is using informers and workqueue to improve its performance over
+// the old controller.
+type CronJobController struct {
+	queue	   			workqueue.RateLimitingInterface
+	recorder   			record.EventRecorder
+
+	jobControl 			jobControlInterface
+	cronJobControl  		cronJobControlInterface
+
+	jobLister  			batchv1listers.JobLister
+	cronJobLister 			CronJobLister
+
+	jobListerSynced 		cache.InformerSynced
+	cronJobListerSynced		cache.InformerSynced
+}
+
+// CronJobControllerConfiguration contains controller config
+type CronJobControllerConfiguration struct {
+	// concurrentCronJobSyncs is the number of cronjob objects that are
+	// allowed to sync concurrently. Larger number = more responsive
+	// but more CPU (and network) load.
+	ConcurrentCronJobSyncs int32
+}
+```
+
+#### Multiple workers
+
+We also propose to have multiple workers controller by a flag similar to [statefulset controller](https://github.com/kubernetes/kubernetes/blob/b58777eda0f0077eba79d063c1c8f31cd7dbb8b9/cmd/kube-controller-manager/app/apps.go#L65). The default would be set to 5 similar to [statefulset](https://github.com/kubernetes/kubernetes/blob/b58777eda0f0077eba79d063c1c8f31cd7dbb8b9/pkg/controller/statefulset/config/v1alpha1/defaults.go#L34)
+
+#### Handling Cron aspect
+
+To detect which CronJob has met its schedule and need to create Jobs we need to implement a timer component. We shall implement a Heap based timer algorithm. We will introduce a separate queue with the [`DelayingInterface`](https://github.com/kubernetes/client-go/blob/b58777eda0f0077eba79d063c1c8f31cd7dbb8b9/util/workqueue/delaying_queue.go#L37) that implements heap based single shot api [`AddAfter`](https://github.com/kubernetes/client-go/blob/b58777eda0f0077eba79d063c1c8f31cd7dbb8b9/util/workqueue/delaying_queue.go#L150). Every time we process an entry from this queue, we will add it back to the queue to simulate a periodic timer.
+
+For further reading:
+1. [Reinventing timer wheel](https://lwn.net/Articles/646950/)
+2. [Hashed and hierarchical timer wheel](http://www.cs.columbia.edu/~nahum/w6998/papers/sosp87-timing-wheels.pdf)
+3. [Golang timers in multi-cpu systems](https://github.com/golang/go/commit/76f4fd8a5251b4f63ea14a3c1e2fe2e78eb74f81)
+4. [Go timerwheel](https://github.com/RussellLuo/timingwheel)
+
+### Metrics
+
+We propose to add metrics that could expose the performance and health of the controller including and not limited to:
+- Skew (actualJobCreationTime-expectedJobCreationTime) - [Histogram](https://prometheus.io/docs/concepts/metric_types/#histogram)
+- Job failures - [Gauge](https://prometheus.io/docs/concepts/metric_types/#counter)
+- Job successes - [Gauge](https://prometheus.io/docs/concepts/metric_types/#counter)
+- API server errors - [Gauge](https://prometheus.io/docs/concepts/metric_types/#counter)
+- Job scheduling latency (podCreationTime - jobCreationTime) - [Histogram](https://prometheus.io/docs/concepts/metric_types/#histogram)
```

Comment:
Shouldn't this be exposed by job controller, instead of cron-job one?

What we can do though maybe, is to label that metric by the "owner kind" (i.e. I'm assuming that Jobs created by CronJob controller will have OwnerRef set to the cronjob objects, and in job controller we can then extract the kind of the owner).

[If your intention was somewhat similar, it probably requires more explantion in the KEP. If it was different, can you explain also in more details what you're planning to do here?]

Reply:
Same as above.